internals-js/src/specs/sourceSpec.ts

```diff
@@ -140,8 +144,271 @@ export class SourceSpecDefinition extends FeatureDefinition {
   sourceFieldDirective(schema: Schema) {
     return this.directive<SourceFieldDirectiveArgs>(schema, 'sourceField')!;
   }
+
+  override validateSubgraphSchema(schema: Schema): GraphQLError[] {
+    const sourceAPIInSchema = schema.directive('sourceAPI') as DirectiveDefinition<SourceAPIDirectiveArgs> | undefined;
+    const sourceTypeInSchema = schema.directive('sourceType') as DirectiveDefinition<SourceTypeDirectiveArgs> | undefined;
+    const sourceFieldInSchema = schema.directive('sourceField') as DirectiveDefinition<SourceFieldDirectiveArgs> | undefined;
+
+    if (!(sourceAPIInSchema || sourceTypeInSchema || sourceFieldInSchema)) {
+      // If none of the @source* directives are present, nothing needs
+      // validating.
+      return [];
+    }
+
+    const errors: GraphQLError[] = [];
+    const apiNameToProtocol = new Map<string, ProtocolName>();
+
+    if (sourceAPIInSchema) {
+      sourceAPIInSchema.applications().forEach(application => {
+        const { name, ...rest } = application.arguments();
+
+        if (apiNameToProtocol.has(name)) {
+          errors.push(new GraphQLError(`${sourceAPIInSchema.name} must specify unique name`));
+        }
+
+        // Ensure name is a valid GraphQL identifier.
+        if (!/^[a-zA-Z_][0-9a-zA-Z_]*$/.test(name)) {
+          errors.push(new GraphQLError(`${sourceAPIInSchema}(name: ${
+            JSON.stringify(name)
+          }) must be valid GraphQL identifier`));
+        }
+
+        let protocol: ProtocolName | undefined;
+        KNOWN_SOURCE_PROTOCOLS.forEach(knownProtocol => {
```

Comment:
Can you not just say `KNOWN_SOURCE_PROTOCOLS.has(protocol)` instead?

Reply:
The basic operation here is to determine which of `KNOWN_SOURCE_PROTOCOLS` is defined as an argument to the `@sourceAPI` directive, which means intersecting `KNOWN_SOURCE_PROTOCOLS` with the directive argument names, so I think iteration is appropriate, since we don't know which protocol will be used in advance (forgetting for now that there's currently only one protocol, `http`).